README.md
Guile DBI
=========

**guile-dbi** provides a simple, generic, easy-to-use **guile scheme**
interface to SQL databases, such as Postgres, MySQL or SQLite3.

The system is 'generic' in the sense that the same programming interface
(front end) can be used with different databases. The DBI (database
independent) part provides the scheme interfaces. The DBD (database
dependent) plugins connect to an actual SQL server. Currently, there are
DBD back-ends for Postgres, MySQL and SQLite3. Creating additional DBD
back-ends requires a small amount of C coding, but is a straightforward
task.

Guile-dbi is simple - which is both a blessing and a curse. For the most
part, all that it does is to accept guile strings encoding SQL
statements, forward these to the database, and return rows as scheme
association lists. A minimal amount of translation is performed - for
example, SQL floating point columns are converted to scheme floating
point numbers - but there is nothing fancier than this - there is no
special treatment for dates, currencies, etc. At this time, there is no
support for prepared statements. Capable programmers are invited to add
support for this and other missing features.

Someone clever might want to figure out how to replace the C code by
generic FFI interfaces, so that there would not be any need at all for
any C code. However, that takes more work, and the code here works OK,
so, for now, things seem OK.

Documentation
-------------
The guile-dbi user-manual and reference is
[here](https://htmlpreview.github.io/?https://github.com/opencog/guile-dbi/blob/master/website/guile-dbi.html).

A copy of the old, defunct website is
[here](https://htmlpreview.github.io/?https://github.com/opencog/guile-dbi/blob/master/website/index.html).

Mailing List
------------
All discussion of guile-dbi should be directed to: guile-user@gnu.org.

Building and Installing
----------------------
Currently, the source code is organized into four distinct projects:
The main database-independent code in DBI, and three different DBD
drivers.  You have to build each independently.  You will need to do
something like this:

```
apt-get install autoconf automake texinfo
```
or maybe
```
yum install autoconf automake texinfo
```
Then:
```
cd guile-dbi
./autogen.sh --no-configure
mkdir build
cd build
../configure
make
sudo make install
```
Then
```
cd guile-dbd-[mysql, postgresql, sqlite3]
./autogen.sh --no-configure
mkdir build
cd build
../configure
make
sudo make install
```

After doing this, the tutorial in
[the user manual](https://htmlpreview.github.io/?https://github.com/opencog/guile-dbi/blob/master/website/guile-dbi.html)
should work fine.


License
-------
Guile-dbi is distributed under the Gnu GPLv2 license. Code and website
were developed by Maurizio Boriani (2005-2006) and are currently
maintained by Linas Vepstas (2008-2019).


guile-dbi/ChangeLog
Version 2.1.9 (2023-07-03):
	* Update autoconf scripts.

Version 2.1.8 (2021-03-13):
	* Fix automake warnings.
	* Enable build even if makeinfo (texinfo) not installed.
	* Remove autogenerated files that should not have been in git.

Version 2.1.7: (2019-09-02)
	* include/guile-dbi/guile-dbi.h (gdbi_db_handle_t): Use 'lt_dlhandle'.
	* src/guile-dbi.c: Replace libdl calls with their libltdl equivalences.
	* Makefile.am (SUBDIRS): Add libltdl.
	* configure.ac: Use LTDL_INIT.  Add LTDLINCL and LIBLTDL to flags.
	* src/Makefile.am (libguile_dbi_la_LIBDD): Remove '-ldl'.
	* src/guile-dbi.c (init_dbi): Call 'lt_dlsetsearchpath'.
	* doc/guile-dbi.texi: Adjust to the usage of libltdl.

Version 2.1.6: (2015-08-01):
	* src/Makefile.am (libguile_dbi_la_LIBADD): Remove '-lz'.
	* src/test/Makefile.am (libguile_dbd_test_la_LIBADD): Remove '-lz'.
	* configure.ac: New option '--with-guile-site-dir' to set 'guilesitedir'.
	* Clear prior dlsym errors before they propagate into guile-dbi.
	* Fix random crashes due to bad interaction with garbage collector.

Version 2.1.5: (2013-04-11):
	* Cleaned up README, TODO files.
	* Missing use of guile-config in test directory.
	* Fix .so vs. .dylib dynalically loadable lib issue for Mac OSX.
	* Remove deprecated guile function warnings.

Version 2.1.4: (2010-05-21):
	* Use guile-config to get library, include-file locations.
	* Add a pkg-config guile-dbi.pc file.

Version 2.1.3: (2010-05-18):
	* Remove deprecated header file guile/gh.h from test case.
	* Addition of sqlite3-dbd from Julian Kalbhenn <jkal@gmx.net> and tantalum@online.de
	* Updated documentation, example usage.

Version 2.1.2 (2008-11-24):
	* Patch from Mike Gran, fixes function names in assert.
	* Minor cleanup; avoid un-needed malloc/free.

2008-09-15 Linas Vepstas <linasvepstas@gmail.com>

	* Numerous bug fixes for crashes, hangs, memory leaks.
	* Bump version to release 2.1.1

2006-09-26  Maurizio Boriani  <baux@gnu.org>

	* DBI-VERIONS: Bump to release 2.0.0.
	* NEWS: Update.
	* configure.ac: Fix libguile check.

2005-07-13  Maurizio Boriani  <baux@gnu.org>

	* DBI-VERIONS: Bump to release 1.0.4.
	* NEWS: Update.

2005-03-21  Maurizio Boriani  <baux@gnu.org>

	* configure.ac: Add output conf file for check-guile-dbi, used by
	greg test framework.
	* DBI-VERSION: Bump release.
	* TODO: Update.

2005-02-19  Maurizio Boriani  <baux@member.fsf.org>

	* README: Add software dependencies.
	* configure.ac: Test for greg suite.

2005-02-11  Maurizio Boriani  <baux@member.fsf.org>

	* Init baux 1.0 arch revision.

2005-02-04  Maurizio Boriani  <baux@member.fsf.org>

	* TODO: Update.
	* Makefile.am: Assign copyright to FSF.
	* README: Add varius info.
	* configure.ac: Test libguile presence.
	* DBI-VERSION: Bump release.
	* NEWS: Update.

2005-01-24  Maurizio Boriani  <baux@member.fsf.org>

	* TODO: Write doc marked done.

2005-01-14  Maurizio Boriani  <baux@member.fsf.org>

	* Makefile.am: Forgotten DBI-VERSION as extra dist file.
	Add doc subdir.
	* DBI-VERSION: Incremented micro version var.
	* configure.ac: Add makefile output for doc.

2005-01-13  Maurizio Boriani  <baux@member.fsf.org>

	* DBI-VERSION: New, versions var
	* TODO: New, thins todo.
	* Makefile.am: Add include subdir.
	Copyright assigner changed to myself
	* configure.ac: Fixed some stuff.
	Copyright assigner changed to myself

2005-01-10  Maurizio Boriani  <baux@member.fsf.org>

	* configure.ac: Add AC_PROG_LIBTOOL.
	* Makefile.am: Fixed typo in banner.

2005-01-08  Maurizio Boriani  <baux@member.fsf.org>

	* doc: New dir for documentation stuff.
	* include: New dir for include files.
	* src: New dir for guile-dbi sources file.
	* COPYING: New, gpl license.
	* AUTHORS: New, authors list.
	* NEWS: New, void.
	* README: New, void.
	* configure.ac: New. Main autoconf stuff.
	* Makefile.am: New. Maim automake makefile.


guile-dbi/src/dbi/dbi.scm
; dbi.scm - scm glue module code
; Copyright (C) 2004, 2005 Free Software Foundation, Inc.
; Written by Maurizio Boriani <baux@member.fsf.org>
;
; This file is part of the guile-dbi.
;
; The guile-dbi is free software; you can redistribute it and/or modify
; it under the terms of the GNU General Public License as published by
; the Free Software Foundation; either version 2 of the License, or
; (at your option) any later version.
;
; The guile-dbi is distributed in the hope that it will be useful,
; but WITHOUT ANY WARRANTY; without even the implied warranty of
; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
; GNU General Public License for more details.
;
; You should have received a copy of the GNU General Public License
; along with this program; if not, write to the Free Software
; Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
;
;                                               -*- Scheme -*-
(define-module (dbi dbi))

(load-extension "libguile-dbi" "init_dbi")

; exported primitives

(export dbi-open
	dbi-close
	dbi-query
	dbi-get_row
	dbi-get_status)


guile-dbi/src/guile-dbi.c
/* guile-dbi.c - Main source file
 * Copyright (C) 2004, 2005, 2006 Free Software Foundation, Inc.
 * Written by Maurizio Boriani <baux@member.fsf.org>
 *
 * This file is part of the guile-dbi.
 * 
 * The guile-dbi is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * The guile-dbi is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 *
 * Process this file with autoconf to produce a configure script. */


#include <guile-dbi/guile-dbi.h>
#include <string.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>


static scm_t_bits g_db_handle_tag;

#define DBI_SMOB_P(obj) ((SCM_NIMP(obj)) && (SCM_TYP16(obj)==g_db_handle_tag))



SCM_DEFINE (make_g_db_handle, "dbi-open", 2, 0, 0,
            (SCM bcknd, SCM conn_string),
            "Build db_handle smob.")
#define FUNC_NAME s_make_g_db_handle
{
  char* sodbd = NULL;
  size_t sodbd_len;
  gdbi_db_handle_t *g_db_handle = NULL;
  void (*connect)(gdbi_db_handle_t*);

  SCM_ASSERT (scm_is_string(bcknd), bcknd, SCM_ARG1, FUNC_NAME);
  SCM_ASSERT (scm_is_string(conn_string), conn_string, SCM_ARG2, FUNC_NAME);

  g_db_handle = (gdbi_db_handle_t *) scm_gc_malloc(sizeof(gdbi_db_handle_t),
                                                   "g_db_handle");

  g_db_handle->bcknd   = bcknd;
  g_db_handle->constr  = conn_string;
  g_db_handle->handle  = NULL;
  g_db_handle->closed  = SCM_BOOL_T;
  g_db_handle->in_free = 0;
  g_db_handle->db_info = NULL;
  g_db_handle->bcknd_str = scm_to_locale_string (bcknd);
  g_db_handle->bcknd_strlen = strlen(g_db_handle->bcknd_str);

  sodbd_len = sizeof(char) * (strlen("libguile-dbd-") +
                   g_db_handle->bcknd_strlen + 1);
  sodbd = (char*) malloc (sodbd_len);
  if (sodbd == NULL)
    {
      g_db_handle->status = scm_cons(scm_from_int(errno),
                                     scm_from_locale_string(strerror(errno)));
      SCM_RETURN_NEWSMOB (g_db_handle_tag, g_db_handle);
    }

  snprintf(sodbd, sodbd_len, "libguile-dbd-%s", g_db_handle->bcknd_str);

  g_db_handle->handle = lt_dlopenext(sodbd);
  if (g_db_handle->handle == NULL)
    {
      free(sodbd);
      g_db_handle->status =  scm_cons(scm_from_int(1),
                              scm_from_locale_string(lt_dlerror()));
      SCM_RETURN_NEWSMOB (g_db_handle_tag, g_db_handle);
    }

  __gdbi_dbd_wrap(g_db_handle, __FUNCTION__, (void**) &connect);
  if (scm_equal_p (SCM_CAR(g_db_handle->status),scm_from_int(0)) == SCM_BOOL_F)
    {
      free(sodbd);
      SCM_RETURN_NEWSMOB (g_db_handle_tag, g_db_handle);
    }
  
  (*connect)(g_db_handle);

  if (sodbd != NULL)
    {
      free(sodbd);
    }

  SCM_RETURN_NEWSMOB (g_db_handle_tag, g_db_handle);
}
#undef FUNC_NAME



static SCM 
mark_db_handle (SCM g_db_handle_smob)
{
  gdbi_db_handle_t* g_db_handle = (gdbi_db_handle_t *)
    SCM_SMOB_DATA(g_db_handle_smob);

  scm_gc_mark(g_db_handle->bcknd);
  scm_gc_mark(g_db_handle->constr);
  scm_gc_mark(g_db_handle->status);
  scm_gc_mark(g_db_handle->closed);

  return SCM_UNSPECIFIED;
}



/* fix this using bcknd function to print */
static int 
print_db_handle (SCM g_db_handle_smob, SCM port,
                 scm_print_state* pstate)
{
  gdbi_db_handle_t* g_db_handle = (gdbi_db_handle_t*)
    SCM_SMOB_DATA(g_db_handle_smob);

  scm_puts("#<guile-dbi ",port);
  if (scm_equal_p (g_db_handle->closed, SCM_BOOL_T) == SCM_BOOL_T)
    {
      scm_puts("close ",port);
    }
  else
    {
      scm_puts("open ",port);
    }
  scm_display(g_db_handle->bcknd,port);
  scm_puts(" ",port);
  scm_display(g_db_handle->constr,port);
  scm_puts(" ",port);
  scm_display(g_db_handle->status,port);
  scm_puts(">",port);
  return 1;
}



SCM_DEFINE (close_g_db_handle, "dbi-close", 1, 0, 0,
            (SCM db_handle),
            "Close db connection.")
#define FUNC_NAME s_close_g_db_handle
{
  gdbi_db_handle_t *g_db_handle = NULL;
  void (*dbd_close)(gdbi_db_handle_t *);

  SCM_ASSERT(DBI_SMOB_P(db_handle), db_handle, SCM_ARG1, FUNC_NAME);
  g_db_handle = (gdbi_db_handle_t *) SCM_SMOB_DATA(db_handle);

  if (g_db_handle->closed == SCM_BOOL_T)
    {
      return  SCM_UNSPECIFIED;
    }

  __gdbi_dbd_wrap(g_db_handle, __FUNCTION__, (void**) &dbd_close);
  if (scm_equal_p (SCM_CAR(g_db_handle->status), scm_from_int(0)) == SCM_BOOL_F)
    {
      scm_remember_upto_here_1(db_handle);
      return  SCM_UNSPECIFIED;
    }

  (*dbd_close)(g_db_handle);
  if (g_db_handle->handle)
    {
      lt_dlclose(g_db_handle->handle);
      g_db_handle->handle = NULL;
    }

  scm_remember_upto_here_1(db_handle);
  return  SCM_UNSPECIFIED;
}
#undef FUNC_NAME



static size_t 
free_db_handle (SCM g_db_handle_smob)
{
  gdbi_db_handle_t *g_db_handle = NULL;

  g_db_handle = (gdbi_db_handle_t *) SCM_SMOB_DATA(g_db_handle_smob);
  if (g_db_handle->in_free) return 0;
  g_db_handle->in_free = 1;

  close_g_db_handle(g_db_handle_smob);

  if (g_db_handle != NULL)
    {
      free((void *) g_db_handle->bcknd_str);
      scm_gc_free(g_db_handle, sizeof(gdbi_db_handle_t), "g_db_handle");
    }

  SCM_SETCDR (g_db_handle_smob, SCM_EOL);
  return 0;
}



SCM_DEFINE (query_g_db_handle, "dbi-query", 2, 0, 0,
            (SCM db_handle, SCM query),
            "Do a query and set status.")
#define FUNC_NAME s_query_g_db_handle
{
  gdbi_db_handle_t *g_db_handle = NULL;
  char             *query_str   = NULL;

  void (*dbi_query)(gdbi_db_handle_t*,char*);

  SCM_ASSERT(DBI_SMOB_P(db_handle), db_handle, SCM_ARG1, FUNC_NAME);  
  SCM_ASSERT(scm_is_string(query), query, SCM_ARG2, FUNC_NAME);

  g_db_handle = (gdbi_db_handle_t *) SCM_SMOB_DATA(db_handle);
  query_str = scm_to_locale_string(query);

  __gdbi_dbd_wrap(g_db_handle, __FUNCTION__, (void**) &dbi_query);
  if (scm_equal_p(SCM_CAR(g_db_handle->status), scm_from_int(0)) == SCM_BOOL_T)
    {
      (*dbi_query)(g_db_handle, query_str);
    }

  free(query_str);
  scm_remember_upto_here_1(db_handle);
  return (SCM_UNSPECIFIED);  
}
#undef FUNC_NAME



SCM_DEFINE (getrow_g_db_handle, "dbi-get_row", 1, 0, 0,
            (SCM db_handle),
            "Do a query and return a row in form of pair list or false")
#define FUNC_NAME s_getrow_g_db_handle
{
  gdbi_db_handle_t *g_db_handle = NULL;
  SCM retrow = SCM_EOL;
  SCM (*dbi_getrow)(gdbi_db_handle_t*);

  SCM_ASSERT (DBI_SMOB_P(db_handle), db_handle, SCM_ARG1, FUNC_NAME);  
  g_db_handle = (gdbi_db_handle_t *) SCM_SMOB_DATA(db_handle);

  __gdbi_dbd_wrap(g_db_handle, __FUNCTION__, (void**) &dbi_getrow);
  if (scm_equal_p (SCM_CAR(g_db_handle->status),scm_from_int(0)) == SCM_BOOL_F)
    {
      scm_remember_upto_here_1(db_handle);
      return(retrow);
    }
  
  retrow = (*dbi_getrow)(g_db_handle);

  scm_remember_upto_here_1(db_handle);
  return(retrow);  
}
#undef FUNC_NAME



SCM_DEFINE (getstat_g_db_handle, "dbi-get_status", 1, 0, 0,
            (SCM db_handle),
            "Return status pair, code and msg, from dbi smob.")
#define FUNC_NAME s_getstat_g_db_handle
{
  gdbi_db_handle_t *g_db_handle = NULL;

  SCM_ASSERT (DBI_SMOB_P(db_handle), db_handle, SCM_ARG1, FUNC_NAME);  

  g_db_handle = (gdbi_db_handle_t *) SCM_SMOB_DATA(db_handle);

  if (g_db_handle != NULL)
    {
      scm_remember_upto_here_1(db_handle);
      return (g_db_handle->status);
    }

  return (SCM_BOOL_F);  
}
#undef FUNC_NAME



/* module init function */
void 
init_db_handle_type(void)
{
  g_db_handle_tag = scm_make_smob_type("g_db_handle",
                                       sizeof(gdbi_db_handle_t));
/* guile-2.0 and newer should not do this.
   See http://lists.gnu.org/archive/html/guile-user/2011-11/msg00074.html */
#if 0 
  scm_set_smob_mark (g_db_handle_tag, mark_db_handle);
#endif
  scm_set_smob_free (g_db_handle_tag, free_db_handle);
  scm_set_smob_print (g_db_handle_tag, print_db_handle);
}


void 
init_dbi(void)
{
  static int is_inited = 0;
  if (is_inited) return;
  is_inited = 1;
  init_db_handle_type();
  lt_dlinit();
  lt_dlsetsearchpath(getenv("GUILE_DBD_PATH"));

#ifndef SCM_MAGIC_SNARFER
#include "guile-dbi.x"
#endif
}



/* dbd handler */
void
__gdbi_dbd_wrap(gdbi_db_handle_t* dbh, const char* function_name,
                void** function_pointer)
{
  const char *ret   = NULL;
  char *func  = NULL;
  size_t func_len;

  func_len = sizeof(char) * (strlen(function_name) + dbh->bcknd_strlen + 10);
  func = malloc(func_len);
  if (NULL == func)
    {
      if (dbh->in_free) return; /* do not SCM anything while in GC */
      dbh->status = scm_cons(scm_from_int(errno),
                             scm_from_locale_string(strerror(errno)));
      return;
    }

  /* I assume this is correct for all OS'es */
  snprintf(func, func_len, "__%s_%s", dbh->bcknd_str, function_name);
  lt_dlerror();    /* Clear any existing error -- Solaris needs this */
  *function_pointer = lt_dlsym(dbh->handle, func);
  if ((ret = lt_dlerror()) != NULL)
    {
      free(func);
      if (dbh->in_free) return; /* do not SCM anything while in GC */
      dbh->status = scm_cons(scm_from_int(1),
                             scm_from_locale_string(ret));
      return;
    }

  free(func);

  if (dbh->in_free) return; /* do not SCM anything while in GC */
  /* todo: error msg to be translated */
  dbh->status = scm_cons(scm_from_int(0),
                         scm_from_locale_string("symbol loaded"));
}


guile-dbi/include/guile-dbi/guile-dbi.h
/* guile-dbi.h - Main include file
 * Copyright (C) 2004, 2005 Free Software Foundation, Inc.
 * Written by Maurizio Boriani <baux@member.fsf.org>
 *
 * This file is part of the guile-dbi.
 *
 * The guile-dbi is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * The guile-dbi is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 */


#ifndef __GUILE_DBI_H__
#define __GUILE_DBI_H__

#include <libguile.h>
#include <ltdl.h>

/* guile smob struct */
typedef struct
{
  SCM bcknd;
  SCM constr;
  SCM status;  /* pair: car = error numeric code, cdr = status message */
  SCM closed;  /* boolean, TRUE if closed otherwise FALSE */
  lt_dlhandle handle;
  void* db_info;
  int in_free;  /* boolean, used to avoid alloc during garbage collection */
  const char * bcknd_str;
  size_t bcknd_strlen;
} gdbi_db_handle_t;
/* end guile smob struct */



/* data wrapper prototypes */
void init_db_handle_type(void);
void init_dbi(void);
/* end data wrapper prototypes */



/* dbd dynamic wrapper stuff */
void __gdbi_dbd_wrap(gdbi_db_handle_t* dbh, const char* function_name,
		     void** function_pointer);
/* end dbd dynamic wrapper stuff */
#endif


guile-dbd-postgresql/src/guile-dbd-postgresql.c
/* guile-dbd-postgresql.c - Main source file
 * Copyright (C) 2004, 2005, 2006,2008 Free Software Foundation, Inc.
 * Written by Maurizio Boriani <baux@member.fsf.org>
 * Maintained by Linas Vepstas <linasvepstas@gmail.com>
 *
 * This file is part of the guile-dbd-postgresql.
 *
 * The guile-dbd-postgresql is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * The guile-dbd-postgresql is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 *
 * */


#include <guile-dbi/guile-dbi.h>
#include <libguile.h>
#include <libpq-fe.h>
#include <string.h>
#include <errno.h>
#include <stdio.h>

/* Function prototypes and structures */
void __postgresql_make_g_db_handle(gdbi_db_handle_t* dbh);
void __postgresql_close_g_db_handle(gdbi_db_handle_t* dbh);
void __postgresql_query_g_db_handle(gdbi_db_handle_t* dbh, char* query);
SCM __postgresql_getrow_g_db_handle(gdbi_db_handle_t* dbh);


typedef struct
{
  PGconn*   pgsql;
  PGresult* res;
  int       lget;
  int       retn;
} gdbi_pgsql_ds_t;



void
__postgresql_make_g_db_handle(gdbi_db_handle_t* dbh)
{
  char sep=':';
  int  items=0;
  gdbi_pgsql_ds_t* pgsqlP = NULL;
  SCM cp_list = SCM_EOL;

  if (scm_equal_p(scm_string_p(dbh->constr), SCM_BOOL_F) == SCM_BOOL_T)
    {
      /* todo: error msg to be translated */
      dbh->status = scm_cons(scm_from_int(1),
                      scm_from_utf8_string("missing connection string"));
      return;
    }

  cp_list = scm_string_split(dbh->constr,SCM_MAKE_CHAR(sep));

  items = scm_to_int(scm_length(cp_list));
  if (items >= 5 && items < 8)
    {
      char* port = NULL;
      char* user, *pass, *db, *ctyp, *loc;

      pgsqlP = (gdbi_pgsql_ds_t*) malloc(sizeof(gdbi_pgsql_ds_t));
      pgsqlP->retn = 0;

      if (pgsqlP == NULL)
        {
          dbh->status = scm_cons(scm_from_int(errno),
                                 scm_from_locale_string(strerror(errno)));
          return;
        }

      user = scm_to_locale_string(scm_list_ref(cp_list,scm_from_int(0)));
      pass = scm_to_locale_string(scm_list_ref(cp_list,scm_from_int(1)));
      db   = scm_to_locale_string(scm_list_ref(cp_list,scm_from_int(2)));
      ctyp = scm_to_locale_string(scm_list_ref(cp_list,scm_from_int(3)));
      loc  = scm_to_locale_string(scm_list_ref(cp_list,scm_from_int(4)));

      pgsqlP->pgsql = NULL;
      pgsqlP->res   = NULL;

      if (strcmp(ctyp,"tcp") == 0)
        {
          port  = scm_to_locale_string(scm_list_ref(cp_list,scm_from_int(5)));
          pgsqlP->pgsql = (PGconn*) PQsetdbLogin(loc,port,NULL,NULL,db,user,pass);
          if (items == 7)
            {
              char* sretn  = scm_to_locale_string(scm_list_ref(cp_list,scm_from_int(6)));
              pgsqlP->retn = atoi(sretn);
              free (sretn);
            }
        }
      else
        {
          pgsqlP->pgsql = (PGconn*) PQsetdbLogin(loc,NULL,NULL,NULL,db,user,pass);
          if (items == 6)
            {
              char* sretn  = scm_to_locale_string(scm_list_ref(cp_list,scm_from_int(5)));
              pgsqlP->retn = atoi(sretn);
              free (sretn);
            }
        }

      /* free resources */
      if (user)
        {
          free(user);
        }
      if (pass)
        {
          free(pass);
        }
      if (db)
        {
          free(db);
        }
      if (ctyp)
        {
          free(ctyp);
        }
      if (loc)
        {
          free(loc);
        }
      if (port)
        {
          free(port);
        }

      if (PQstatus(pgsqlP->pgsql) == CONNECTION_BAD)
        {
          dbh->status = scm_cons(scm_from_int(1),
                     scm_from_locale_string(PQerrorMessage(pgsqlP->pgsql)));
          PQfinish(pgsqlP->pgsql);
          pgsqlP->pgsql = NULL;
          free(pgsqlP);
          dbh->db_info = NULL;
          dbh->closed = SCM_BOOL_T;
          return;
        }
      else
        {
          /* todo: error msg to be translated */
          dbh->status = scm_cons(scm_from_int(0),
                          scm_from_utf8_string("db connected"));
          dbh->db_info = pgsqlP;
          dbh->closed = SCM_BOOL_F;
          return;
        }
    }
  else
    {
      /* todo: error msg to be translated */
      dbh->status = scm_cons(scm_from_int(1),
                     scm_from_utf8_string("invalid connection string"));
      dbh->db_info = NULL;
      dbh->closed = SCM_BOOL_T;
      return;
    }

  return;
}



void
__postgresql_close_g_db_handle(gdbi_db_handle_t* dbh)
{
  gdbi_pgsql_ds_t* pgsqlP = (gdbi_pgsql_ds_t*)dbh->db_info;
  if (pgsqlP == NULL)
    {
      if (dbh->in_free) return; /* don't scm anything if in GC */
      /* todo: error msg to be translated */
      dbh->status = scm_cons(scm_from_int(1),
                 scm_from_utf8_string("dbd info not found"));
      return;
    }

  /* Don't scm anything if we are in garbage collection */
  if (0 == dbh->in_free)
    {
      if (NULL == pgsqlP->pgsql)
        {
          /* todo: error msg to be translated */
          dbh->status = scm_cons(scm_from_int(1),
                 scm_from_utf8_string("dbi connection already closed"));
        }
      else
        {
          dbh->status = scm_cons(scm_from_int(0),
                      scm_from_utf8_string("dbi closed"));
        }
    }

  PQfinish(pgsqlP->pgsql);
  if (pgsqlP->res)
    {
      PQclear(pgsqlP->res);
      pgsqlP->res = NULL;
    }

  free(dbh->db_info);
  dbh->db_info = NULL;

  dbh->closed = SCM_BOOL_T;
}



void
__postgresql_query_g_db_handle(gdbi_db_handle_t* dbh, char* query)
{
  gdbi_pgsql_ds_t* pgsqlP = NULL;
  int err, i, bpid;

  if (dbh->db_info == NULL)
    {
      /* todo: error msg to be translated */
      dbh->status = scm_cons(scm_from_int(1),
                   scm_from_utf8_string("invalid dbi connection"));
      return;
    }

  if (query == NULL)
    {
      /* todo: error msg to be translated */
      dbh->status = scm_cons(scm_from_int(1),
                 scm_from_utf8_string("invalid dbi query"));
      return;
    }

  pgsqlP = (gdbi_pgsql_ds_t*)dbh->db_info;

  /* read up all results before next query */

  do
    {
      if (pgsqlP->res)
        {
          PQclear(pgsqlP->res);
          pgsqlP->res = NULL;
        }
    }
  while ((pgsqlP->res = PQgetResult(pgsqlP->pgsql)) != NULL); 

#if 0
  if (PQresultStatus(pgsqlP->res) == PGRES_FATAL_ERROR)
    {
      for (i = 0, bpid=0; i < pgsqlP->retn && bpid == 0; i++)
        {
          PQreset(pgsqlP->pgsql);
          bpid = PQbackendPID(pgsqlP->pgsql);
        }
    }
#endif
  err = PQsendQuery(pgsqlP->pgsql, query);

  if (err == 1)
    {
      dbh->status = scm_cons(scm_from_int(0),
                    scm_from_utf8_string("query ok"));
      pgsqlP->lget = 0;
    }
  else
    {
      dbh->status = scm_cons(scm_from_int(1),
              scm_from_locale_string(PQerrorMessage(pgsqlP->pgsql)));
    }
}



SCM
__postgresql_getrow_g_db_handle(gdbi_db_handle_t* dbh)
{
  gdbi_pgsql_ds_t* pgsqlP = NULL;
  SCM retrow = SCM_EOL;
  int fnum,f ;

  if (dbh->db_info == NULL)
    {
      /* todo: error msg to be translated */
      dbh->status = scm_cons(scm_from_int(1),
                   scm_from_utf8_string("invalid dbi connection"));
      return (SCM_BOOL_F);
    }

  pgsqlP = (gdbi_pgsql_ds_t*)dbh->db_info;
  if (NULL == pgsqlP->res)
      pgsqlP->res = PQgetResult(pgsqlP->pgsql);

  if (NULL == pgsqlP->res)
    {
      dbh->status = scm_cons(scm_from_int(0),
                             scm_from_utf8_string("row end"));
      pgsqlP->lget = 0;
      return SCM_BOOL_F;
    }

  if (pgsqlP->lget == PQntuples(pgsqlP->res))
    {
      pgsqlP->lget = 0;
      PQclear(pgsqlP->res);
      pgsqlP->res = PQgetResult(pgsqlP->pgsql);
    }

  /* Check result status before unpacking the data! */
  switch (PQresultStatus(pgsqlP->res))
    {
    case PGRES_BAD_RESPONSE:
    case PGRES_NONFATAL_ERROR:
    case PGRES_FATAL_ERROR:
      if (pgsqlP->res == NULL)
        {
          dbh->status = scm_cons(scm_from_int(0),
                                scm_from_utf8_string("row end"));
        }
      else
        {
          dbh->status = scm_cons(scm_from_int(1),
               scm_from_locale_string(PQresStatus(PQresultStatus(pgsqlP->res))));
        }
      return SCM_BOOL_F;

    case PGRES_EMPTY_QUERY:
    case PGRES_COMMAND_OK:
    case PGRES_TUPLES_OK:
    case PGRES_COPY_OUT:
    case PGRES_COPY_IN:
      dbh->status = scm_cons(scm_from_int(0),
            scm_from_locale_string(PQresStatus(PQresultStatus(pgsqlP->res))));
      break;

    default:
      dbh->status = scm_cons(scm_from_int(0),
            scm_from_utf8_string("unknown return query status"));
      break;
    }

  fnum = PQnfields(pgsqlP->res);
  for (f = 0; f<fnum; f++)
    {
      SCM value;

      /* The different field types can be gotten by saying
       * SELECT typname, oid from pg_type;
       * They do not seem to be listed in any header files... */
      Oid type = PQftype(pgsqlP->res, f);
      if ((type >= 20 && type <= 24) || /* int2, int4, int8 */
          type == 26                  ) /* oid */
        {
          const char * vstr = PQgetvalue(pgsqlP->res, pgsqlP->lget,f);
          value = scm_from_long_long(atoll(vstr));
        }
      else if (type == 700 || /* float4 */
               type == 1700 || /* numeric */
               type == 701 )  /* float8 */
        {
          const char * vstr = PQgetvalue(pgsqlP->res, pgsqlP->lget, f);
          if (vstr && *vstr)
            { 
              value = scm_c_locale_stringn_to_number(vstr, strlen(vstr), 10);
            }
          else
            {
              value = scm_from_double(0.0);
            }
        }
      else if (type == 1005 || /* _int2  -- list of integers */
               type == 1007 || /* _int4  -- list of integers */
               type == 1016 )  /* _int8  -- list of integers */
        {
          const char * vstr = PQgetvalue(pgsqlP->res, pgsqlP->lget, f);
          char * p = strdup(vstr + 1);
          char * tok = rindex(p, ',');
          value = SCM_EOL;
          while (tok)
            {
              SCM vtok = scm_from_long_long(atoll(tok + 1));
              value = scm_cons(vtok, value);
              *tok = 0x0;
              tok = rindex(p, ',');
            }
          SCM vtok = scm_from_long_long(atoll(p));
          value = scm_cons(vtok, value);
          free(p);
        }
      else if (type == 18 ||  /* char */
               type == 19 ||  /* name */
               type == 25 ||  /* text */
               type == 702 || /* abstime !!! XXX */
               (type >= 1042 && type <= 1114)) /* varchar, timestamps */
        {
          const char * vstr = PQgetvalue(pgsqlP->res, pgsqlP->lget, f);
          value = scm_from_locale_string(vstr);
        }
      else
        {
          char msg[101];
          snprintf(msg, 100, "unknown field type %d for %s", type,
                   PQfname(pgsqlP->res, f));
          dbh->status = scm_cons(scm_from_int(1),
                 scm_from_utf8_string(msg));
          pgsqlP->lget++;
          return SCM_EOL;
        }

      retrow = scm_append(scm_list_2(retrow,
          scm_list_1(scm_cons(scm_from_locale_string(PQfname(pgsqlP->res,f)),
                                              value))));
    }

  pgsqlP->lget++;

  if (retrow == SCM_EOL) retrow = SCM_BOOL_F;
  return retrow;
}
